docs/spec/distributing.rst

```diff
@@ -10,72 +10,428 @@ Stub files
 
-Stub files are files containing type hints that are only for use by
-the type checker, not at runtime.  There are several use cases for
-stub files:
+*Stub files*, also called *type stubs*, provide type information for untyped
+Python packages and modules. Stub files serve multiple purposes:
+
+* They are the only way to add type information to extension modules.
+* They can provide type information for packages that do not wish to
+  add them inline.
+* They can be distributed separately from the implementation.
+  This allows stubs to be developed at a different pace or by different
+  authors, which is especially useful when adding type annotations to
+  existing packages.
+* They can act as documentation, succinctly explaining the external
+  API of a package, without including the implementation or private
+  members.
+
+Stub files that only use the constructs described in :ref:`stub-file-supported-constructs`
+below should work with all type checkers that conform to this specification. A
+conformant type checker will parse a stub that only uses such constructs without
+error and will not interpret any construct in a contradictory manner. However,
+type checkers are not required to implement checks for all these constructs and
+can elect to ignore unsupported ones. Additionally, type checkers can support
+constructs not decribed here, and tool authors are encouraged to experiment with
+additional features.
+
+.. _stub-file-syntax:
+
+Syntax
+^^^^^^
+
+Stub files are syntactically valid Python files in the earliest Python version
+that is not yet end-of-life. They use a ``.pyi`` suffix. The Python syntax used
+by stub files is independent from the Python versions supported by the
+implementation, and from the Python version the type checker runs under (if
+any). Therefore, stub authors should use the latest syntax features available in
+the earliest supported version, even if the implementation supports older
+versions. Type checker authors are encouraged to support syntax features from
+newer versions, although stub authors should not use such features if they wish
+to maintain compatibility with all type checkers.
+
+For example, Python 3.7 added the ``async`` keyword (see :pep:`492`).
+Stub authors should use it to mark coroutines, even if the implementation
+still uses the ``@coroutine`` decorator. On the other hand, stubs should not use
+the ``type`` soft keyword from :pep:`695`, introduced in Python 3.12, util
+Python 3.11 reaches end-of-life in October 2027.
+
+Stubs are treated as if ``from __future__ import annotations`` is enabled. In
+particular, built-in generics, pipe union syntax (``X | Y``), and forward
+references can be used.
+
+Type checkers should have a configurable search path for stub files. If a stub
+file is found, the type checker should not read the corresponding "real" module.
+See :ref:`mro` for more information.
+
+.. _stub-file-supported-constructs:
+
+Supported Constructs
```

Comment:
I find the current contents of this section to be a bit confusing because it vacillates between "here's what a conformant type checker must support" and "here's guidance for a stub author". I think the typing spec should include only the former. The latter should be included in guides.

Reply:
Thanks for this feedback! I've redrafted this section to focus on what type checkers should support and moved any stub author guidance that seemed useful to keep to the writing stubs guide.